### ISSUE TITLE
[C#] Implement new Concept Exercise: parameters

### ISSUE THREAD

**Issue by ErikSchierboom** [exercism/v3#1128]
This issue describes how to implement the `parameters` concept exercise for the C# track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [The features of v3](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md).
- [Rationale for v3](https://github.com/exercism/v3/blob/master/docs/rationale-for-v3.md).
- [What are concept exercise and how they are structured?](https://github.com/exercism/v3/blob/master/docs/features-of-v3.md)

Please also watch the following video:

- [The Anatomy of a Concept Exercise](https://www.youtube.com/watch?v=gkbBqd7hPrA).

## Goal

The goal of this exercise is to teach the student the Concept of Parameters in C#.

## Learning objectives

- Know the difference between value and reference type parameters.
- Know how to pass value types by reference using the `ref`, `out` and `in` modifiers.

## Out of scope

- Overload resolution.
- ref returns.
- ref locals.

## Concepts

This Concepts Exercise's Concepts are:

- `parameters`: know the difference between value and reference type parameters; know how to pass value types by reference using the `ref`, `out` and `in` modifiers.

## Prequisites

This Concept Exercise's prerequisites Concepts are:

- `memory-allocation`: know what value and reference types are and how they relate to memory usage, the heap and the stack.

Any data types used in this exercise (e.g. `strings`) should also be added as prerequisites.

### Hints

- [passing-parameters][passing-parameters]: explains how values can be passed as arguments.
- [ref-parameter][ref-parameter]: describes how `ref` parameters work.
- [out-parameter][out-parameter]: describes how `out` parameters work.
- [in-parameter][in-parameter]: describes how `in` parameters work.

### After

- [passing-parameters][passing-parameters]: explains how values can be passed as arguments.
- [ref-parameter][ref-parameter]: describes how `ref` parameters work.
- [out-parameter][out-parameter]: describes how `out` parameters work.
- [in-parameter][in-parameter]: describes how `in` parameters work.

## Representer

This exercise does not require any specific representation logic to be added to the [representer][representer].

## Analyzer

This exercise does not require any specific analyzer logic to be added to the [analyzer][analyzer].

## Implementing

To implement this exercise, please [follow these instructions](https://github.com/exercism/v3/blob/master/languages/csharp/reference/implementing-a-concept-exercise.md).

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue.

[how-to-implement-a-concept-exercise]: https://github.com/exercism/v3/blob/master/docs/maintainers/generic-how-to-implement-a-concept-exercise.md
[implemented-exercises]: https://github.com/exercism/v3/tree/master/languages/csharp/exercises/concept/README.md#implemented-exercises
[reference]: https://github.com/exercism/v3/blob/master/languages/csharp/reference/README.md#reference-docs
[reference-example]: https://github.com/exercism/v3/blob/master/reference/types/string.md#implementations
[analyzer]: https://github.com/exercism/csharp-analyzer
[representer]: https://github.com/exercism/csharp-representer
[exercise-example]: https://github.com/exercism/v3/tree/master/languages/csharp/exercises/concept/numbers-floating-point
[design-example]: https://github.com/exercism/v3/blob/master/languages/csharp/exercises/concept/numbers/.meta/design.md
[config.json-example]: https://github.com/exercism/v3/blob/master/languages/csharp/exercises/concept/numbers/.meta/config.json
[concept-exercises]: https://github.com/exercism/v3/blob/master/docs/concept-exercises.md
[passing-parameters]: https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-references/passing-parameters
[ref-parameter]: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/ref#passing-an-argument-by-reference
[in-parameter]: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/in-parameter-modifier
[out-parameter]: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/out-parameter-modifier

**Comment by mikedamay** [comment]
I am thinking of enhancing the remote controlled cars exercise.

- `out` parameters for fuel usage and tire wear
- `varargs` for sponsor names
- `in` parameter to move to `structs` where it makes more sense.
- `ref` parameter awaits inspiration.

**Comment by ErikSchierboom** [comment]
I am thinking of enhancing the remote controlled cars exercise.

> `out` parameters for fuel usage and tire wear

How do you envision this. Having a `GetStatus` method or something like that?

> `varargs` for sponsor names

I like this!

> `in` parameter to move to structs where it makes more sense.

Good use case. So this would be doing something with a struct but not modifying it.

> `ref` parameter awaits inspiration.

Maybe have a method that gets passed a struct but _does_ modify it? To contrast it with the previous `in` parameter case?

**Comment by mikedamay** [comment]
> a GetStatus method or something like that?

Yep.

> Maybe have a method that gets passed a struct but does modify it? To contrast it with the previous in parameter case?

I did not really want to take s`truct`s as a dependency.  `ref` makes sense with scalars.  `in` pretty much needs `structs` to make sense and we can't enforce it with tests.

**Comment by ErikSchierboom** [comment]
> I did not really want to take structs as a dependency. ref makes sense with scalars. in pretty much needs structs to make sense and we can't enforce it with tests.

I wouldn't mind only having the `in` modifier in the `after.md` file only. From my experience, I've not seen it being used often. You?

**Comment by mikedamay** [comment]
>  From my experience, I've not seen it being used often. You?

Never.  As you say _after.md_ is a good place for `in`.  Maybe mentioned here and then elaborated in the _after.md_ for `structs`.

So, unless you have any further thoughts, I will go ahead with `out`, `varargs` and `ref` which will probably clunk a bit.  I may as well include default values (although they are introduced in the `methods` exercise and I suppose `named parameters`.

**Comment by ErikSchierboom** [comment]
Named parameters are actually introduced in the `basics` exercise: https://github.com/exercism/v3/blob/master/languages/csharp/exercises/concept/basics/.docs/introduction.md But I'd say including them here too definitely makes sense.

**Comment by mikedamay** [comment]
Closed by PR #1699
